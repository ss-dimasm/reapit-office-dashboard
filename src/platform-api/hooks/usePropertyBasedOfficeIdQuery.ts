import { ReapitConnectSession } from '@reapit/connect-session'
import { PropertyModelPagedResult } from '@reapit/foundations-ts-definitions'
import axios from 'axios'
import { useQuery } from 'react-query'

import { BASE_HEADERS, URLS } from '../../constants/api'

const getPropertiesByOfficeId = async (
  session: ReapitConnectSession,
  officeId: string
): Promise<PropertyModelPagedResult | undefined> => {
  if (!session) return

  const { data } = await axios.get(
    `${window.reapit.config.platformApiUrl}${URLS.PROPERTIES_PAGED}?officeId=${officeId}&pageSize=100`,
    {
      headers: {
        ...BASE_HEADERS,
        Authorization: `Bearer ${session.accessToken}`,
      },
    }
  )
  return data
}

const usePropertyBasedOfficeIdQuery = (
  session: ReapitConnectSession,
  officeId: string
) => {
  return useQuery(
    ['properties', officeId],
    () => getPropertiesByOfficeId(session, officeId),
    {
      enabled: !!session,
    }
  )
}

export default usePropertyBasedOfficeIdQuery
