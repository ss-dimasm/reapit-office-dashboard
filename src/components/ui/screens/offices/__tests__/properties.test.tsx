import React from 'react'

import { Subtitle, Table } from '@reapit/elements'
import { mount } from 'enzyme'
import { QueryClient, QueryClientProvider } from 'react-query'

import usePropertyBasedOfficeIdQuery from '../../../../../platform-api/hooks/usePropertyBasedOfficeIdQuery'
import LoadingIndicator from '../../../loading'
import Properties from '../properties'
import {
  PROPERTIES_SCREEN_EXAMPLE_1,
  PROPERTIES_SCREEN_EXAMPLE_2,
} from '../__mocks__/properties.mock'

const mockUserPropertyBasedOfficeIdQuery =
  usePropertyBasedOfficeIdQuery as jest.Mock<any>
jest.mock('../../../../../platform-api/hooks/usePropertyBasedOfficeIdQuery')

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
})
describe('<Properties/>', () => {
  beforeEach(() => {
    mockUserPropertyBasedOfficeIdQuery.mockImplementation(() => ({
      isLoading: true,
      data: PROPERTIES_SCREEN_EXAMPLE_1,
    }))
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should render <Properties/>', () => {
    const wrapper = mount(
      <QueryClientProvider client={queryClient}>
        <Properties officeData={jest.fn()} />
      </QueryClientProvider>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should show loading indicator', () => {
    const wrapper = mount(
      <QueryClientProvider client={queryClient}>
        <Properties officeData={jest.fn()} />
      </QueryClientProvider>
    )
    expect(wrapper.find(LoadingIndicator).text()).toEqual('Please wait...')
  })

  it('should show notification while data is 0', () => {
    mockUserPropertyBasedOfficeIdQuery.mockImplementation(() => ({
      isLoading: false,
      data: PROPERTIES_SCREEN_EXAMPLE_2,
    }))
    const wrapper = mount(
      <QueryClientProvider client={queryClient}>
        <Properties officeData={jest.fn()} />
      </QueryClientProvider>
    )
    const notification = wrapper.find(Subtitle)
    expect(notification.text()).toEqual('No properties handled')
    expect(notification).toHaveLength(1)
  })

  it('should render table of properties', () => {
    mockUserPropertyBasedOfficeIdQuery.mockImplementation(() => ({
      isLoading: false,
      data: PROPERTIES_SCREEN_EXAMPLE_1,
    }))
    const wrapper = mount(
      <QueryClientProvider client={queryClient}>
        <Properties officeData={jest.fn()} />
      </QueryClientProvider>
    )

    const tableProperties = wrapper.find(Table)
    expect(tableProperties.text()).toContain('Test Building Name')
    expect(tableProperties).toHaveLength(1)
  })
})
