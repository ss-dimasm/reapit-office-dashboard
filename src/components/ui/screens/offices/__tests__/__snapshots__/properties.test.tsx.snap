// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Properties/> should render <Properties/> 1`] = `
<QueryClientProvider
  client={
    QueryClient {
      "defaultOptions": Object {
        "queries": Object {
          "retry": false,
        },
      },
      "mutationCache": MutationCache {
        "config": Object {},
        "listeners": Array [],
        "mutationId": 0,
        "mutations": Array [],
      },
      "mutationDefaults": Array [],
      "queryCache": QueryCache {
        "config": Object {},
        "listeners": Array [],
        "queries": Array [],
        "queriesMap": Object {},
      },
      "queryDefaults": Array [],
      "unsubscribeFocus": [Function],
      "unsubscribeOnline": [Function],
    }
  }
>
  <Properties
    officeData={[MockFunction]}
  >
    <LoadingIndicator
      isFullScreen={true}
      label="Please wait..."
    >
      <div
        className="el-flex el-flex-justify-center elH-full el-flex-align-center"
        style={
          Object {
            "height": "70%",
          }
        }
      >
        <Ut
          data-testid="loadingIndicator"
          label="Please wait..."
        >
          <ElLoaderContainer
            className=""
          >
            <div>
              <ElLoaderLabel>
                <p>
                  Please wait...
                </p>
              </ElLoaderLabel>
              <ElLoader
                data-testid="loadingIndicator"
              >
                <div
                  data-testid="loadingIndicator"
                >
                  <ElLoaderMovingBar>
                    <div />
                  </ElLoaderMovingBar>
                </div>
              </ElLoader>
            </div>
          </ElLoaderContainer>
        </Ut>
      </div>
    </LoadingIndicator>
  </Properties>
</QueryClientProvider>
`;
